File: dotnet/source/amp.rabbit/transport/MultiConnectionRabbitSender.cs
using System;
using System.Collections;
using System.Collections.Generic;
using amp.rabbit.connection;
using amp.rabbit.topology;
using cmf.bus;
using Common.Logging;
using RabbitMQ.Client;

namespace amp.rabbit.transport
{
    public class MultiConnectionRabbitSender : IDisposable
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(MultiConnectionRabbitSender));

        protected IConnectionManagerCache _connectionManagerCache;

        public MultiConnectionRabbitSender(IConnectionManagerCache connectionManagerCache)
        {
            _connectionManagerCache = connectionManagerCache;
        }

        public void Send(RoutingInfo routing, Envelope envelope)
        {
            Log.Debug("Enter Send");

            // for each exchange, send the envelope
            foreach (ProducingRoute route in routing.ProducingRoutes)
            {
                Log.Info("Sending to exchange: " + route.Exchange.ToString());

                //TODO: There is a problem here in how exceptions are being logged/rethrown:
                //It is possible for a envelope to go out on some exchanges and/or routes and then not others.
                //A failure aborts attempts at subequent exchanges/routes but does not role back previous.
                //Seems we should either continue to attempt later routes and throw at the end or role back.
                try {
                    IEnumerable<IConnectionManager> connectionManagers = _connectionManagerCache.GetConnectionManagersFor(route);

                    foreach (var connectionManager in connectionManagers)
                    {
                        using (IModel channel = connectionManager.CreateModel())
                        {
                            SendDataOnChannel(channel, route, envelope);
                        }
                    }

                } catch (Exception e) {
                    Log.Error("Failed to send an envelope", e);
                    throw;
                }
            }

            Log.Debug("Leave Send");        
        }

        protected static void SendDataOnChannel(IModel channel, ProducingRoute route, Envelope envelope)
        {
            IBasicProperties props = channel.CreateBasicProperties();
            props.Headers = envelope.Headers as IDictionary;

            Exchange ex = route.Exchange;
            if (ex.ShouldDeclare)
            {
                channel.ExchangeDeclare(ex.Name, ex.ExchangeType, ex.IsDurable, ex.IsAutoDelete, ex.ArgumentsAsDictionary);
            }

            foreach (string routingKey in route.RoutingKeys)
            {
                try
                {
                    channel.BasicPublish(ex.Name, routingKey, props, envelope.Payload);
                }
                catch (Exception e)
                {
                    Log.Error(
                        string.Format("Failed to send an envelope to route: {0} on excange {1}.",
                            routingKey, ex), e);
                    throw;
                }
            }

            // close the channel, but not the connection.  Channels are cheap.
            channel.Close();
        }

        public void Dispose()
        {
            _connectionManagerCache.Dispose();
        }
    }
}

File: java/amp-rabbit/src/main/java/amp/rabbit/transport/MultiConnectionRabbitSender.java
package amp.rabbit.transport;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import cmf.bus.Envelope;
import cmf.bus.IDisposable;

import com.rabbitmq.client.AMQP;
import com.rabbitmq.client.Channel;

import amp.rabbit.connection.IConnectionManager;
import amp.rabbit.connection.IConnectionManagerCache;
import amp.rabbit.topology.Exchange;
import amp.rabbit.topology.ProducingRoute;
import amp.rabbit.topology.RoutingInfo;


/**
 * This abstracts the publishing of information to multiple channels & brokers,
 * managing the connections and fanout of the distribution.
 * 
 * @author jmccune
 */
public class MultiConnectionRabbitSender implements IDisposable  {

	private static final Logger LOG = LoggerFactory.getLogger(MultiConnectionRabbitSender.class);
	
	/** The factory that provides us our connections & channels */
	private IConnectionManagerCache _connectionFactory;
	
	public MultiConnectionRabbitSender(IConnectionManagerCache connectionFactory) {
		_connectionFactory = connectionFactory;
	}
	
	public void send(RoutingInfo routing, Envelope envelope)  throws Exception {
		LOG.debug("Enter Send");
		
		// for each exchange, send the envelope
		for (ProducingRoute route : routing.getProducingRoutes()) {
		
			try {
				
				Collection<IConnectionManager> connectionManagers = _connectionFactory.getConnectionManagersFor(route);
						
				for (IConnectionManager mgr : connectionManagers) {
					Channel channel 	 = mgr.createChannel();
					sendDataOnChannel(channel,route, envelope);
				}			
			} catch (Exception e) {
				LOG.error("Failed to send an envelope", e);
				throw e;
			}
		}

		LOG.debug("Leave Send");

	}
	
	protected void sendDataOnChannel(Channel channel,ProducingRoute route,  Envelope envelope) {
		
		try { 
			AMQP.BasicProperties props = new AMQP.BasicProperties.Builder().build();

			//Convert Headers from String,String to String,Object Map.
			Map<String, Object> headers = new HashMap<String, Object>();
			for (Map.Entry<String, String> entry : envelope.getHeaders()
					.entrySet()) {
				headers.put(entry.getKey(), entry.getValue());
			}

			//Set the properties
			props.setHeaders(headers);

			Exchange ex = route.getExchange();
			if(ex.shouldDeclare()){
				channel.exchangeDeclare(ex.getName(), ex.getExchangeType(),
						ex.isDurable(), ex.isAutoDelete(), ex.getArguments());
			}

			// For the channel, publish the topic...
			for (String key : route.getRoutingKeys()) {
				try {
					channel.basicPublish(ex.getName(), key, props, envelope.getPayload());
				} catch (Exception e) {
					LOG.error("Failed to send an envelope to route: " + key);
							
					throw e;
				}
			}
		} catch (Exception x) {
			LOG.error("Error attempting to send envelope: ", x);
		}
	}

	@Override
	public void dispose() {
		//Shoudl I dispose of hte factory...??
		
		try { _connectionFactory.dispose(); } 
		catch(Exception x) { LOG.warn("Error closing connection factory cache.", x);}
	}
}